File: gatewayapi/src/main/java/com/example/gatewayapi/exception/ApiException.java
package com.example.gatewayapi.exception;

class ApiException extends RuntimeException {
    private final ExceptionCode code;

    public ApiException(ExceptionCode code) {
        super(code.getMessage());
        this.code = code;
    }

    public ApiException(ExceptionCode code, String message) {
        super(message);
        this.code = code;
    }

    public ExceptionCode getCode() {
        return code;
    }
}

File: services/common/src/main/kotlin/com/example/common/exception/ApiException.kt
package com.example.common.exception

class ApiException(val code: ApiExceptionCode, private val customMessage: String? = null): Exception() {
    override val message: String
        get() = customMessage ?: code.defaultMessage
}

enum class ApiExceptionCode(val defaultMessage: String = "") {

    // Database related
    DATABASE_RECORD_NOT_FOUND("We could not find the specified record"),
    DATABASE_ACCESS_ERROR("Unknown Database Error"),
    DATABASE_MORE_THAN_ONE_RECORD("Found more than one record"),

    // Data quality related
    VALIDATION_ERROR("Model has invalid data"),

    // Core related
    USER_NOT_FOUND("User not found"),

    // 3rd Party API related
    EXTERNAL_API_CONNECTION_ERROR(),
    EXTERNAL_API_DATA_CONVERSION_ERROR("Could not interpret the API response"),
    EXTERNAL_API_OTHER("External API returned an error"),
}